### DISCUSSION TITLE
Harmonize timestamps of an existing calendar

### DISCUSSION THREAD

**Discussion by marsianer** [ical4j/ical4j#644] [Q&A]
I have a calendar that uses different time zones. 
For this reason I would like to _harmonize_ all entries to use the same time zone _Europe/Berlin_.

- Most of them are floating and UTC (Legacy of GMail).
- I want to make sure that the existing entries don't get broken.

Is this coding correct or did I miss something?

```
CompatibilityHints.setHintEnabled(CompatibilityHints.KEY_RELAXED_PARSING, true);
for (CalendarComponent calendarComponent : getCalendar().getComponents()) {
    // Start date
    Property dtstartProperty = calendarComponent.getProperty(Property.DTSTART);
    if (dtstartProperty != null) {
        DtStart startDate = (DtStart) dtstartProperty;
        if (startDate.getDate() instanceof DateTime) {
            startDate.setTimeZone(getDefaultTimeZone());
        }
    }
    // End date
    Property dtendProperty = calendarComponent.getProperty(Property.DTEND);
    if (dtendProperty != null) {
        DtEnd endDate = (DtEnd) dtendProperty;
        if (endDate.getDate() instanceof DateTime) {
            endDate.setTimeZone(getDefaultTimeZone());
        }
    }
    // Due date
    Property dueProperty = calendarComponent.getProperty(Property.DUE);
    if (dueProperty != null) {
        Due dueDate = (Due) dueProperty;
        dueDate.setTimeZone(getDefaultTimeZone());
    }
    // Exclusion date
    Property exdateProperty = calendarComponent.getProperty(Property.EXDATE);
    if (exdateProperty != null) {
        ExDate exclusionDate = (ExDate) exdateProperty;
        exclusionDate.setTimeZone(getDefaultTimeZone());
    }
    // Recurrence date
    Property rdateProperty = calendarComponent.getProperty(Property.RDATE);
    if (rdateProperty != null) {
        RDate recurrenceDate = (RDate) rdateProperty;
        recurrenceDate.setTimeZone(getDefaultTimeZone());
    }
}
// Save to file...
```

**Comment by benfortuna** [comment]
First thing that comes to mind is recurrence instances. If you have any recurring events, and have overridden any of them you may have RDATE properties which will also need a TZID reference.

You may want to look at the spec (RFC5545) just to see which properties can have a TZID parameter, and cross-check with your file. It is an interesting project and probably could be made into a general tool for this type of conversion..

**Comment by marsianer** [comment]
Time Zone Identifier
[https://datatracker.ietf.org/doc/html/rfc5545#section-3.2.19](url)

This parameter MUST be specified on
- DTSTART (Start date)
- DTEND (End date)
- DUE (Due date)
- EXDATE (Exclusion date)
- RDATE (Recurrence date)

when either a DATE-TIME or TIME value type is specified and when the value is neither a UTC or a "floating" time.

**Comment by marsianer** [comment]
Since I set the time zone explicitly on the properties mentioned, do I still need the VTIMEZONE entries?

**Reply by benfortuna** [reply]
Strictly yes, as the TZID parameter is meant to be a reference to a VTIMEZONE component.

But you may find some implementations will allow you to use well known timezone identifiers without needing to include the VTIMEZONE definition (e.g. Europe/London).

**Comment by marsianer** [comment]
Thanks for your help.
The harmonization has just been successfully completed.

All events now use the same time zone.
